bsms.mediawiki

```diff
@@ -0,0 +1,181 @@
+
+The Signer is a participating member in the multisig. Its responsibilities include providing its key record -- which contains an Extended Public Key (XPUB) -- to the Coordinator, verifying that its XPUB is included in the descriptor record and persisting the descriptor record in its storage.
+
+===Setup Process===
+
+====Round 1====
+
+=====Coordinator=====
+
+* The Coordinator creates a multisig wallet creation session. The Coordinator constructs the multisig script and its policy parameters, such as the total number of signers and the required number of signatures (<tt>M</tt> and <tt>N</tt>).
+* The session should expire after some time period determined by the Coordinator, e.g., 24 hours. The timeout allows the encryption key to have lower entropy.
+* If encryption is enabled, the Coordinator distributes a secret <tt>TOKEN</tt> to each Signer over a secure channel. The Signer can use the <tt>TOKEN</tt> to derive an <tt>ENCRYPTION_KEY</tt>. Refer to the Encryption section below for details on the <tt>TOKEN</tt>, the key derivation function and the encryption scheme. Depending on the use case, the Coordinator can decide whether to share one common <tt>TOKEN</tt> for all Signers, or to have one per Signer.
+* If encryption is disabled, the <tt>TOKEN</tt> is set to <tt>0</tt>, and all the encryption/decryption steps below can be skipped.
+
+=====Signer=====
+
+* The Signer initiates a new secure multisig setup session by setting the <tt>TOKEN</tt>. The Signer derives an <tt>ENCRYPTION_KEY</tt> from the <tt>TOKEN</tt>. The Signer can keep the session open until a different value for the <tt>TOKEN</tt> is set.
+* The Signer generates a key record by prompting the user for a multisig derivation path and retrieves the XPUB at that derivation path. Optionally, the Signer can choose a path on behalf of the user. If the Signer chooses the path, it should try to avoid reusing XPUBs for different wallets.
+* The first line in the record must be the hex-encoded <tt>TOKEN</tt>. The second line must be the <tt>KEY</tt>. The <tt>KEY</tt> is an XPUB plus its key origin information, written in the descriptor-defined format, i.e.: <tt>[{master key fingerprint}/{derivation path}]{XPUB}</tt>. The third line is a text description of the key, 80 characters maximum. The fourth line must be a <tt>SIG</tt>, whereas <tt>SIG</tt> is the signature generated by using the private key associated with the XPUB to sign the first three lines. The signature should follow [https://github.com/bitcoin/bips/blob/master/bip-0322.mediawiki BIP-0322], legacy format accepted.
+* The Signer calculates the Message Authentication Code (<tt>MAC</tt>) for the record. The <tt>MAC</tt> doubles as the Initialization Vector (<tt>IV</tt>) for the encryption.
+* The Signer encrypts the key record with the <tt>ENCRYPTION_KEY</tt> and <tt>IV</tt>.
+* The Signer encodes the <tt>MAC</tt> and the ciphertext into hexadecimal format, then concatenates the results: <tt>(MAC || ciphertext)</tt>.
```

Comment:
I favour hex-encodnig over base64 encoding for its simplicity and being easier in debugging - you can just see the byte values directly, without the need to convert. It is also straightforward to concatenate hex-encoded values, whereas with base64 I think it is better to concatenate the binary data first, and then encode.

On the other hand, hex encoding is wasteful in terms of the size of the resulting data.

Xpub is 111 chars in lengh, also up to 80 chars of text description, plus the token, plus 32 bytes of MAC. Text description length may vary, so let's say that we have about 200 bytes of the resulting encrypted data with MAC. That's 268 chars in base64, or 400 chars in hex.

The question is if this difference of ~130 bytes is significant.

If these records are to be transferred via QR codes, for example, it might make sense to conserve space. 268 chars vs 400 chars might mean using 57x57 qr code vs 65x65 qr code (https://www.qrcode.com/en/about/version.html)

We discussed the matter off-PR with @hugohn but we both do not have strong opinion on either option (base64 or hex), and it would be great to have more people to weight in. Allowing both base64 and hex will make the spec unnecessary complex, IMO.